lib/galaxy/tools/parameters/dynamic_options.py

```diff
@@ -319,13 +319,17 @@ def get_dependency_name(self):
         return self.dynamic_option.dataset_ref_name
 
     def filter_options(self, options, trans, other_values):
-        rval = []
-        skip_list = []
-        for fields in options:
-            if fields[self.column] not in skip_list:
-                rval.append(fields)
-                skip_list.append(fields[self.column])
-        return rval
+        seen = set()
+        dup_count, length = 0, len(options)
+        for ix, fields in enumerate(options):
+            if fields[self.column] in seen:
+                dup_count += 1
+                continue
+            seen.add(fields[self.column])
+            if dup_count:
+                options[ix - dup_count] = fields
+        options[length - dup_count : length] = []
+        return options
```

Comment:
```suggestion
        rval = []
        skip_list = set()
        for fields in options:
            if fields[self.column] not in skip_list:
                rval.append(fields)
                skip_list.add(fields[self.column])
        return rval
```
This is a clear improvement for large lists. I am wary to change the options in place, that could have unintended consequences. At worst this scales with the number of options, and is probably not worth optimizing, and it makes the code harder to read.

Reply:
I agree with @mvdbeek .. just for completeness: A correct and simpler version of an in place replacement would have been

```
def filter_options(options, column):
    seen = set()
    for ix, fields in enumerate(options):
        field = fields[column]
        if field in seen:
            continue
        options[len(seen)] = fields
        seen.add(field)
    del options[len(seen):]
    return options
```

In particular I think the `if dup_count:` line was wrong (because this would lead to skipping the first entry).

Reply:
@bernt-matthias  Your code looks cleaner. I like it. I think there is no wrong with the dup_count though. Assuming bool checker is lighter than reassigning the element, we could skip changing the list as many times as possible. Also, the first element can be safely skipped.

Reply:
I can restore the returning in a new list, but keep the list to set change.